### PR TITLE
Fix for RNN/LSTM/GRU modules to work with stateless.functional_call()

### LINKED ISSUE #90500: nn.LSTM not working together with functional_call for calculating the gradient
### 🐛 Describe the bug

I have the following code:

```
import torch
from torch.nn.utils.stateless import functional_call
import torch.autograd as autograd
import torch.nn as nn

# This is the model
class Encoder(nn.Module):
    def __init__(self, action_dim, z_dim, skill_length):
        super().__init__()
        print(action_dim)
        self.lin1 = nn.Linear(action_dim, action_dim)
        self.lstm = nn.LSTM(input_size=action_dim, hidden_size=z_dim, batch_first=True)
        self.lin2 = nn.Linear(z_dim, z_dim)

    def forward(self, skill):
        a, b, c = skill.shape
        skill = skill.reshape(-1, skill.shape[-1])
        embed = self.lin1(skill)
        embed = embed.reshape(a, b, c)
        mean, _ = self.lstm(embed)
        mean = mean[:, -1, :]
        mean = self.lin2(mean)

        return mean

# This is the initialization function
def pars(model):
    params = {}
    for name, param in model.named_parameters():
        if len(param.shape) == 1:
            init = torch.nn.init.constant_(param, 0)
        else:
            init = torch.nn.init.orthogonal_(param)
        params[name] = nn.Parameter(init)
    return params

# Initializating the model
model = Encoder(4, 2, 5)
x = torch.rand(3, 5, 4)
params = pars(model)


# Running the model with functional_call and calculating gradient.
samp = functional_call(model, params, x)
grad_f = autograd.grad(torch.mean(samp), params.values(),
                       retain_graph=True, allow_unused=True)
print(grad_f)
# grad_f has gradient for the linear layer, but None for the LSTM layer.

# Running the model without functional_call and calculating gradient.
samp = model(x)
grad = autograd.grad(torch.mean(samp), model.parameters(), retain_graph=True)
print(grad)
# grad has gradient for all layers, e.g., linears and lstm.
```

I know the problem is with the LSTM layer because when I use a linear layer with ```nn.Linear```, then the gradient depends on std as well as the linear layer. Unfortunately, I do not know to resolve this problem. I'd appreciate any help.

*Edit: I heavily edited the code provided just to further simplify the example. This code can be copied and run.

**Update Dec 11, 2022**

```
class Encoder(nn.Module):
    def __init__(self, action_dim, z_dim, skill_length):
        super().__init__()
        print(action_dim)
        self.lin1 = nn.Linear(action_dim, action_dim)
        self.lstm = nn.LSTM(input_size=action_dim, hidden_size=z_dim, batch_first=True)
        self.lin2 = nn.Linear(z_dim, z_dim)

    def forward(self, skill):
        a, b, c = skill.shape
        skill = skill.reshape(-1, skill.shape[-1])
        embed = self.lin1(skill)
        embed = embed.reshape(a, b, c)
        mean, _ = self.lstm(embed)
        pdb.set_trace()
        grad1 = autograd.grad(mean.mean(), params.values(),
                              retain_graph=True, allow_unused=True)
        # This gives gradient for the self.lin1 layer, and None for the LSTM
        grad2 = autograd.grad(mean.mean(), self.parameters(),
                              retain_graph=True, allow_unused=True)
        # This gives gradient the LSTM, but None for the self.lin1 layer
        mean = mean[:, -1, :]
        mean = self.lin2(mean)

        return mean
```

When I run it the regular without `functional_call` and calling directly the model, then `autograd.grad(mean.mean(), self.parameters(), allow_unused=True, retain_graph=True)` has gradient for the self.lin1 and LSTM layer.

I don't know if this information is useful, but putting out there just in case.

### Versions

--2022-12-08 16:03:15--  https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 17278 (17K) [text/plain]
Saving to: ‘collect_env.py’

collect_env.py      100%[===================>]  16.87K  --.-KB/s    in 0s

2022-12-08 16:03:15 (108 MB/s) - ‘collect_env.py’ saved [17278/17278]

Collecting environment information...
PyTorch version: 1.13.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Fedora Linux 36 (Workstation Edition) (x86_64)
GCC version: (GCC) 12.2.1 20220819 (Red Hat 12.2.1-2)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.8 (main, Oct 12 2022, 00:00:00) [GCC 12.2.1 20220819 (Red Hat 12.2.1-2)] (64-bit runtime)
Python platform: Linux-6.0.7-200.fc36.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 520.56.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.0
[pip3] torchaudio==0.13.0
[pip3] torchvision==0.14.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ngimel

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91111

Fixes #90500

The change here checks for parameter changes at the beginning of each `forward()` call; if the parameters are found to be different tensors than last time, `self._flat_weights` is re-initialized with the new values. Thus, swapping parameter values using `stateless.functional_call()` will re-initialize `self._flat_weights` during the `forward()` call, and the provided parameters will be used for module computation as expected.

NB: There are still some changes needed for symbolic shapes to work with `nn.GRU` (will address in a follow-up PR).

cc @albanD @mruberry @walterddr @kshitij12345 @saketh-are

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
ahhh, this is very good, thank you!

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -258,6 +274,12 @@ def extra_repr(self) -> str:
             s += ', bidirectional={bidirectional}'
         return s.format(**self.__dict__)
 
+    def __getstate__(self):
+        # Don't serialize the weight references.
+        state = self.__dict__.copy()
```
Comment (albanD) [inline torch/nn/modules/rnn.py:279]: should this be `state = super().__getstate__()` ?

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -268,33 +290,32 @@ def __setstate__(self, d):
...
-                        self._all_weights += [weights[:2]]
-                        self._flat_weights_names.extend(weights[:2])
-        self._flat_weights = [(lambda wn: getattr(self, wn) if hasattr(self, wn) else None)(wn) for wn in self._flat_weights_names]
+                        if self.proj_size > 0:
+                            self._all_weights += [weights[:2]] + [weights[-1:]]
+                            self._flat_weights_names.extend(weights[:2] + [weights[-1:]])
+                        else:
+                            self._all_weights += [weights[:2]]
+                            self._flat_weights_names.extend(weights[:2])
+        self._init_flat_weights()
```
Comment (albanD) [inline torch/nn/modules/rnn.py:318]: The old code for this function was not calling `self.flatten_parameters()`?

File: torch/nn/modules/rnn.py [hunk 3]
```diff
@@ -123,11 +125,16 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
...
                 self._all_weights.append(param_names)
 
-        self._flat_weights = [(lambda wn: getattr(self, wn) if hasattr(self, wn) else None)(wn) for wn in self._flat_weights_names]
-        self.flatten_parameters()
+        self._init_flat_weights()
 
         self.reset_parameters()
 
+    def _init_flat_weights(self):
+        self._flat_weights = [(lambda wn: getattr(self, wn) if hasattr(self, wn) else None)(wn) for wn in self._flat_weights_names]
```
Comment (albanD) [inline torch/nn/modules/rnn.py:133]: Not yours but:
```
self._flat_weights = [getattr(self, wn, None) for wn in self._flat_weights_names]
```

**Review by albanD — APPROVED** [review]
Looks good!
Only small nits

File: torch/nn/modules/rnn.py (hunk 3 above)
Comment (jbschlosser) [reply torch/nn/modules/rnn.py:133]: Haha good call, super weird to use a lambda this way

File: torch/nn/modules/rnn.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nn/modules/rnn.py:279]: I tried that first, but apparently `super().__getstate__()` didn't exist until Python 3.11 for some reason.

AFAICT from the python docs, the default behavior when a `__getstate__()` impl is not present is equivalent to returning the object's `state_dict`.

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (jbschlosser) [reply torch/nn/modules/rnn.py:318]: You're correct about this. It seemed incomplete to me since the other two places this is done do include a call to `flatten_parameters()`, but I'm happy to revert it back just in case.

File: torch/nn/modules/rnn.py (hunk 3 above)
Comment (jbschlosser) [reply torch/nn/modules/rnn.py]: So mypy does not like the use of `getattr(self, wn, None)`. i can still get rid of the lambda though
